bin/src/cli.rs

```diff
@@ -80,28 +81,30 @@ pub fn get_upgrade_fd<'a>(matches: &ArgMatches<'a>) -> i32 {
     .parse::<i32>().expect("the file descriptor must be a number")
 }
 
-pub fn upgrade_worker<'a>(matches: &ArgMatches<'a>) -> Option<()> {
-  matches.subcommand_matches("worker").and_then(|worker_matches| {
+pub fn upgrade_worker<'a>(matches: &ArgMatches<'a>) -> Result<Option<()>, anyhow::Error> {
+  if let Some(worker_matches) = matches.subcommand_matches("worker") {
     let fd = get_fd(worker_matches);
     let scm = get_scm(worker_matches);
     let configuration_state_fd = get_configuration_state_fd(worker_matches);
     let id = get_id(worker_matches);
     let buffer_size = get_buffer_size(worker_matches);
     let max_buffer_size = get_max_buffer_size(worker_matches, buffer_size);
 
-    begin_worker_process(fd, scm, configuration_state_fd, id, buffer_size, max_buffer_size);
-    Some(())
-  })
+    Ok(Some(begin_worker_process(fd, scm, configuration_state_fd, id, buffer_size, max_buffer_size)?))
+  } else {
+    Ok(None)
+  }
 }
 
-pub fn upgrade_main<'a>(matches: &ArgMatches<'a>) -> Option<()> {
- matches.subcommand_matches("upgrade").and_then(|upgrade_matches| {
+pub fn upgrade_main<'a>(matches: &ArgMatches<'a>) -> Result<Option<()>, anyhow::Error> {
```

Comment:
I think, you should change the signature of the function to Result<(), anyhow::Error>. To me the Option was here to handle the error without return it

Reply:
I would agree, but actually the Option is usefull in `main.rs`:

```rust
// Check if we are upgrading workers
  let mut upgrade = cli::upgrade_worker(&matches)?;

  // check if we are upgrading main (overrides worker upgrading)
  if let Some(upgrade_main) = cli::upgrade_main(&matches)? {
    upgrade = Some(upgrade_main);
  }
            
  // If we are not, then we want to start sozu
  if upgrade == None {
```

Reply:
Which we could refactor as well, I'm all for it.

Reply:
yup, it should get a bit of refactoring, the code is not very clear. I think at some point the option was holding some data, but I'm not sure